vms/platformvm/txs/executor/state_changes.go

```diff
@@ -93,17 +111,21 @@ func AdvanceTimeTo(
 
 	changes, err := state.NewDiffOn(parentState)
 	if err != nil {
-		return false, err
+		return nil, false, err
 	}
 
+	// Promote any pending stakers to current if [StartTime] <= [newChainTime].
+	//
+	// Invariant: It is not safe to modify the state while iterating over it,
+	// so we use the parentState's iterator rather than the changes iterator.
+	// ParentState must not be modified before this iterator is released.
 	pendingStakerIterator, err := parentState.GetPendingStakerIterator()
 	if err != nil {
-		return false, err
+		return nil, false, err
 	}
 	defer pendingStakerIterator.Release()
```

Comment:
This was perilously close to being an existing race.

The iterator is released in a `defer`, but the changes are applied on the `return` line. The `defer` is executed _after_ the `return` line is evaluated. Which previously broke the invariant (that is now documented). However, because we only ever passed in a `state.Diff` into this function and the `state.Diff` lazily allocates the `btree`'s that are iterated over, the returned iterator wasn't a tree iterator, but just explicitly empty... Which _does_ support writes during iteration.

The `expiry` tree is allocated during diff initialization, which is why this function showed a race when adding the expiry iterator.

By moving the state application to the parent caller, the `defer` executes before `Apply` and there is no chance of a race.